beacon_node/rest_api/src/helpers.rs

```diff
@@ -169,6 +170,33 @@ pub fn implementation_pending_response(_req: Request<Body>) -> ApiResult {
     ))
 }
 
+pub fn get_beacon_chain_from_request<T: BeaconChainTypes + 'static>(
+    req: &Request<Body>,
+) -> Result<(Arc<BeaconChain<T>>, BeaconState<T::EthSpec>), ApiError> {
+    // Get beacon state
+    let beacon_chain = req
+        .extensions()
+        .get::<Arc<BeaconChain<T>>>()
+        .ok_or_else(|| ApiError::ServerError("Beacon chain extension missing".into()))?;
+    let mut head_state = beacon_chain
```

Comment:
Regarding the final paragraph in my previous message:

We actually ran into a full beacon chain dead-lock yesterday due to taking long-lasting read locks on the head. As such, I've made it so you can't get a read-lock on the head from the `BecaonChain` anymore. Now whenever you call `beacon_chain.head()` you get a _clone_ of the head instead of a read-lock to it. The changes to downstream functions are quite minimal.

Of course, this is more copying but it protects us from dead-locks.

Reply:
Good to know!
I have addressed this and I only get the head state when it's necessary.